src/channel.cr

```diff
@@ -127,23 +275,50 @@ abstract class Channel(T)
     self.select ops
   end
 
-  def self.select(ops : Tuple | Array, has_else = false)
-    loop do
-      ops.each_with_index do |op, index|
-        if op.ready?
-          result = op.execute
-          return index, result
-        end
-      end
+  def self.select(ops : Indexable(SelectAction), has_else = false)
+    # Sort the operations by the channel they contain
+    # This is to avoid deadlocks between concurrent `select` calls
+    ops_locks = ops
+      .to_a
+      .uniq(&.lock_object_id)
+      .sort_by(&.lock_object_id)
```

Comment:
Considering that `Channel.select` will often be called in a loop, allocations inside might end up costly if the actual work to handle each op is cheap. This might perhaps be microoptimized for `ops` with few entries along the same lines asterite optimized small Hashes recently?

(To hyperoptimize it, it might even make sense to create some sort of presorted registry that keep them in order that allow registrying or de-registrying selectables. But that might be overkill outside  #6468 :) ).